src/business_logic/state/update_tracker_state.rs

```diff
@@ -0,0 +1,139 @@
+    // value to storage that is identical to the value previously held at that address, then no
+    // change is made to that cell and it does not count as a storage-change in fee calculation.
+
+    pub fn set_storage_at(
+        &mut self,
+        contract_address: Felt,
+        key: u64,
+        value: Felt,
+    ) -> Result<(), StateError> {
+        let address_key_pair = (contract_address, key);
+        if !self.was_accessed(&address_key_pair) {
+            let val = self.state.get_storage_at(&address_key_pair)?;
+            let value = val
+                .to_u64()
+                .ok_or_else(|| StateError::ConversionError(value.clone()))?;
+            self.storage_initial_values
+                .insert(address_key_pair.clone(), value);
+        }
+
+        let stor_value = value
+            .to_u64()
+            .ok_or_else(|| StateError::ConversionError(value.clone()))?;
+        self.storage_writes
+            .insert(address_key_pair.clone(), stor_value);
+        self.state.set_storage_at(&address_key_pair, value);
+        Ok(())
+    }
+
+    pub fn get_storage_at(&mut self, contract_address: Felt, key: u64) -> Result<Felt, StateError> {
+        let address_key_pair = (contract_address.clone(), key);
+        let was_accessed = !self.was_accessed(&address_key_pair);
+
+        let return_value = self.state.get_storage_at(&(contract_address, key))?;
+
+        if was_accessed {
+            let value = return_value
+                .to_u64()
+                .ok_or_else(|| StateError::ConversionError(return_value.clone()))?;
+            self.storage_initial_values.insert(address_key_pair, value);
+        }
+        Ok(return_value.clone())
+    }
+
+    pub fn block_info(&self) -> &BlockInfo {
+        self.state.block_info()
+    }
+
+    pub fn update_block_info(&mut self, mut block_info: BlockInfo) {
+        self.state.update_block_info(block_info)
+    }
+
+    pub fn get_contract_class(&mut self, class_hash: &[u8]) -> Result<&ContractClass, StateError> {
+        self.state.get_contract_class(class_hash)
+    }
+
+    pub fn get_nonce_at(&mut self, contract_address: &Felt) -> Result<&Felt, StateError> {
+        self.state.get_nonce_at(contract_address)
+    }
+
+    pub fn set_contract_class(&mut self, class_hash: &[u8], contract_class: &ContractClass) {
+        self.state.set_contract_class(class_hash, contract_class)
+    }
+
+    pub fn deploy_contract(
+        &mut self,
+        contract_address: Felt,
+        class_hash: Vec<u8>,
+    ) -> Result<(), StateError> {
+        self.state.deploy_contract(contract_address, class_hash)
+    }
+
+    pub fn increment_nonce(&mut self, contract_address: &Felt) -> Result<(), StateError> {
+        self.state.increment_nonce(contract_address)
+    }
+
+    pub fn count_actual_storage_changes(&self) -> (usize, usize) {
+        let storage_updates = self
+            .storage_writes
+            .clone()
```

Comment:
Why not just `into_iter`?

Reply:
Because it says that it was behind a shared reference and we would loose the ownership of `self` as far a i understand, which is not the intended behavior.